### PR TITLE
Tellurics, Sky lines and correction to Spectroscopic lines

### PR DESCRIPTION
**Pull request by rjoebruch**
Hi! 
I modified the plot.py code to add skylines and tellurics bands, which are not dependant on the redshift or expansion velocity parameters. 

I also modified one error in the O III lines (all lines are forbidden lines and should only be [O III]. 

Let me know if there's something else to add here. 

Best! 

--

### CODE REVIEW

File: skyportal/plot.py [hunk 1]
```diff
@@ -1472,20 +1577,80 @@ def spectroscopy_plot(obj_id, user, spec_id=None, width=600, device="browser"):
...
+                x=midtel2,
+                width=widtel2,
+                # TODO change limits
+                top=1e4,
+                color=color,
+                alpha=0.3,
+                # source=ColumnDataSource(tel2),
+            )
+
+            sz_dict[f'el{i}'].visible = False
```
Comment (dmitryduev) [inline skyportal/plot.py:1638]: Looks almost identical, how about:

```suggestion
        elif name_line_list[i][0] in ('Tellurics-1', 'Tellurics-2'):

            el_data['x'] = el_data['wavelength']
            midtel1 = (el_data['x'][0] + el_data['x'][1]) / 2
            widtel1 = el_data['x'][1] - el_data['x'][0]

            sz_dict[f'el{i}'] = plot.vbar(
                x=midtel1,
                width=widtel1,
                # TODO change limits
                top=1e4,
                color=color,
                alpha=0.3,
            )

            sz_dict[f'el{i}'].visible = False
```

File: skyportal/plot.py [hunk 2]
```diff
@@ -1498,6 +1663,10 @@ def spectroscopy_plot(obj_id, user, spec_id=None, width=600, device="browser"):
 
     elements_groups = []  # The Bokeh checkbox groups
     callbacks = []  # The checkbox callbacks for each element
+
+    full_model_dict = {**model_dict, **sz_dict}
+    # dict3 = {**dict1 , **dict2}
```
Comment (dmitryduev) [inline skyportal/plot.py:1668]:
```suggestion
```

File: skyportal/plot.py [hunk 3]
```diff
@@ -1472,20 +1577,80 @@ def spectroscopy_plot(obj_id, user, spec_id=None, width=600, device="browser"):
...
+            widtel1 = el_data['x'][1] - el_data['x'][0]
+
+            sz_dict[f'el{i}'] = plot.vbar(
+                x=midtel1,
+                width=widtel1,
+                # TODO change limits
+                top=1e4,
+                color=color,
+                alpha=0.3,
+                # source=ColumnDataSource(tel1),
```
Comment (dmitryduev) [inline skyportal/plot.py:1615]:
```suggestion
```

File: skyportal/plot.py [hunk 4]
```diff
@@ -1512,7 +1681,34 @@ def spectroscopy_plot(obj_id, user, spec_id=None, width=600, device="browser"):
                 'elements': elements,
                 'z': z_textinput,
                 'v_exp': v_exp_textinput,
-                **model_dict,
+                # **model_dict,
```
Comment (dmitryduev) [inline skyportal/plot.py:1684]:
```suggestion
```

File: skyportal/plot.py [hunk 5]
```diff
@@ -1566,7 +1765,11 @@ def spectroscopy_plot(obj_id, user, spec_id=None, width=600, device="browser"):
     )
 
     # Update the element spectral lines as well
-    for callback in callbacks:
+    for (
+        callback
+    ) in (
+        callbacks
+    ):  # it's all the callbacks and it's wroth for every one of them. should separate here also
```
Comment (dmitryduev) [inline skyportal/plot.py:1772]:
```suggestion
    for callback in callbacks:  
        # it's all the callbacks and it's wroth for every one of them. should separate here also
```

**Review by dmitryduev — APPROVED** [review]
LGTM, a few minor comments. CI Failure seems unrelated.

**Comment by rjoebruch** [comment]
Thanks for the suggestions, much more ergonomic, thanks!

**Comment by rjoebruch** [comment]
Is there something to add /check to finalise this?

**Comment by dmitryduev** [comment]
@rjoebruch could you please pull in the latest SP and see if the CI is happy?

File: skyportal/plot.py [hunk 6]
```diff
@@ -1552,8 +1698,37 @@ def spectroscopy_plot(obj_id, user, spec_id=None, width=600, device="browser"):
...
+        callback_zvs = CustomJS(
+            args={
+                'elements': elements,
+                'z': z_textinput,
+                'v_exp': v_exp_textinput,
+                # **model_dict,
+                **model_dict,  # only on the spec lines, not tellurics or sky lines
+            },
+            code=f"""
+            let c = 299792.458; // speed of light in km / s
```
Comment (acrellin) [inline skyportal/plot.py:1712]:
```suggestion
            const c = 299792.458; // speed of light in km / s
```

File: skyportal/plot.py [hunk 7]
```diff
@@ -1588,7 +1763,8 @@ def spectroscopy_plot(obj_id, user, spec_id=None, width=600, device="browser"):
     )
 
     # Update the element spectral lines as well
-    for callback in callbacks:
+    for callback in callbacks:  
+        # it's all the callbacks and it's wroth for every one of them. should separate here also
```
Comment (acrellin) [inline skyportal/plot.py:1767]: Not sure what this comment is supposed to mean.

**Review by acrellin — APPROVED** [review]
Looks good, a couple minor comments below. Thanks @rjoebruch !

File: skyportal/plot.py (hunk 7 above)
Comment (rjoebruch) [reply skyportal/plot.py:1767]: it was for my own understanding, can be removed

File: skyportal/plot.py [hunk 8]
```diff
@@ -1494,20 +1599,58 @@ def spectroscopy_plot(obj_id, user, spec_id=None, width=600, device="browser"):
         margin=(0, 10, 0, 10),
     )
 
+    name_line_list = list(SPEC_LINES.items())
+
     for i, (wavelengths, color) in enumerate(SPEC_LINES.values()):
```
Comment (kmshin1397) [inline skyportal/plot.py:1604]:
```suggestion
    for i, (name, (wavelengths, color)) in enumerate(SPEC_LINES.items()):
```

File: skyportal/plot.py [hunk 9]
```diff
@@ -1494,20 +1599,58 @@ def spectroscopy_plot(obj_id, user, spec_id=None, width=600, device="browser"):
         margin=(0, 10, 0, 10),
     )
 
+    name_line_list = list(SPEC_LINES.items())
+
```
Comment (kmshin1397) [inline skyportal/plot.py:1603]:
```suggestion
```

File: skyportal/plot.py [hunk 10]
```diff
@@ -1494,20 +1599,58 @@ def spectroscopy_plot(obj_id, user, spec_id=None, width=600, device="browser"):
...
-            x1='x',
-            # TODO change limits
-            y0=0,
-            y1=1e4,
-            color=color,
-            source=ColumnDataSource(el_data),
-        )
-        model_dict[f'el{i}'].visible = False
+
+        if name_line_list[i][0] == 'Sky Lines':  # No redshift correction of skylines
```
Comment (kmshin1397) [inline skyportal/plot.py:1608]:
```suggestion
        if name == 'Sky Lines':  # No redshift correction of skylines
```

File: skyportal/plot.py [hunk 11]
```diff
@@ -1494,20 +1599,58 @@ def spectroscopy_plot(obj_id, user, spec_id=None, width=600, device="browser"):
...
+                x1='x',
+                y0=0,
+                y1=1e4,
+                color=color,
+                source=ColumnDataSource(el_data),
+            )
+
+            sz_dict[f'el{i}'].visible = False
+
+        elif name_line_list[i][0] in ('Tellurics-1', 'Tellurics-2'):
```
Comment (kmshin1397) [inline skyportal/plot.py:1623]:
```suggestion
        elif name in ('Tellurics-1', 'Tellurics-2'):
```

**Review by kmshin1397 — COMMENTED** [review]
Thanks @rjoebruch this looks good! My suggested changes are more for cosmetic purposes - your code isn't wrong but I think this will make it a little easier to follow :)

Also, looks like the CI tests are failing to run due to linting issues. If you haven't yet, please install our pre-commit hooks as shown [here](https://skyportal.io/docs/dev.html#pre-commit-hook) and then try running `pre-commit run --from-ref ${The latest commit hash of master} --to-ref ${Your PR's commit hash}` to fix the linting issues. Thanks!

**Comment by stefanv** [comment]
I will incorporate the review feedback and push to this branch.  I found a small problem with the callbacks and will address that too.

**Comment by stefanv** [comment]
Please let me know if this is the desired behavior.

skylines

**Review by dmitryduev — APPROVED** [review]
@stefanv: thanks for fixing this PR up!
Yes, this is the expected behavior. Tellurics and sky lines originate from the Earth's atmosphere; there are tons of them, so it depends how sensitive a survey is, and what band they are observing in. With those two sliders you are Doppler-shifting the source.

**Pull request merged**